src/System.Private.CoreLib/shared/System/Diagnostics/Debug.cs

```diff
@@ -342,9 +375,46 @@ private sealed class DebugAssertException : Exception
             }
         }
 
+        private static void ShowDialog(string stackTrace, string message, string detailMessage, string errorSource)
+        {
+            if (s_provider.AssertUIEnabled)
+            {
+                s_provider.ShowDialog(stackTrace, message, detailMessage, errorSource);
+            }
+            else if (Debugger.IsAttached)
+            {
+                Debugger.Break();
+            }
+            else
+            {
+                // In Core, we do not show a dialog.
+                // Fail in order to avoid anyone catching an exception and masking
+                // an assert failure.
+                DebugAssertException ex;
+                if (message == string.Empty) 
+                {
+                    ex = new DebugAssertException(stackTrace);
+                }
+                else if (detailMessage == string.Empty) 
+                {
+                    ex = new DebugAssertException(message, stackTrace);
+                }
+                else
+                {
+                    ex = new DebugAssertException(message, detailMessage, stackTrace);
+                }
+                Environment.FailFast(ex.Message, ex, errorSource);
+            }
+        }
+
         // internal and not readonly so that the tests can swap this out.
         internal static Action<string, string, string, string> s_ShowDialog = ShowDialog;
```

Comment:
In the tests I noticed they switch back to default when the test is done. I would need to have a mechanism to set back to default then. 
Would it make sense for `DebugProvider Debug.s_provider` to be `DebugProvider[] Debug.s_providers` instead and I would have an extra API on Debug called SetPrimary(DebugProvider) that helps switch implementations?

Reply:
> `DebugProvider[]`

I do not think this low-level hook should deal with arrays of providers. 

You can solve this by changing `RegisterProvider` to return the previous provider. Also, maybe a better name for `RegisterProvider` would be `SetProvider` - since the method does not let you register more than one.